python_modules/dagster/dagster/core/execution/context/output.py

```diff
@@ -372,6 +384,116 @@ def get_output_identifier(self) -> List[str]:
+
+        Events logged with this method will appear in the list of DagsterEvents, as well as the event log.
+
+        Args:
+            event (Union[AssetMaterialization, Materialization, AssetObservation]): The event to log.
+
+        Examples:
+
+        .. code-block:: python
+            from dagster import IOManager, AssetMaterialization
+
+            class MyIOManager(IOManager):
+                def handle_output(self, context, obj):
+                    context.log_event(AssetMaterialization("foo"))
+        """
+        from dagster.core.events import DagsterEvent
+
+        if isinstance(event, (AssetMaterialization, Materialization)):
+            if self._step_context:
+                self._events.append(
+                    DagsterEvent.asset_materialization(
+                        self._step_context,
+                        event,
+                        self._step_context.get_input_lineage(),
+                    )
+                )
+            self._user_events.append(event)
+        elif isinstance(event, AssetObservation):
+            if self._step_context:
+                self._events.append(DagsterEvent.asset_observation(self._step_context, event))
+            self._user_events.append(event)
+        else:
+            check.failed("Unexpected event {event}".format(event=event))
+
+    def consume_events(self) -> Iterator["DagsterEvent"]:
+        """Pops and yields all user-generated events that have been recorded from this context.
+
+        If consume_events has not yet been called, this will yield all logged events since the call to `handle_output`. If consume_events has been called, it will yield all events since the last time consume_events was called. Designed for internal use. Users should never need to invoke this method.
+        """
+
+        events = self._events
+        self._events = []
+        yield from events
+
+    def get_logged_events(
+        self,
+    ) -> List[Union[AssetMaterialization, Materialization, AssetObservation]]:
+        """Retrieve the list of user-generated events that were logged via the context.
+
+
+        User-generated events that were yielded will not appear in this list.
+
+        **Examples:**
+
+        .. code-block:: python
+
+            from dagster import IOManager, build_output_context, AssetMaterialization
+
+            class MyIOManager(IOManager):
+                def handle_output(self, context, obj):
+                    ...
+
+            def test_handle_output():
+                mgr = MyIOManager()
+                context = build_output_context()
+                mgr.handle_output(context)
+                all_user_events = context.get_logged_events()
+                materializations = [event for event in all_user_events if isinstance(event, AssetMaterialization)]
+                ...
+        """
+
+        return self._user_events
+
+    def add_metadata_entry(
```

Comment:
Any reason to make this only accept a single metadata entry, while the parallel method on the op context accepts a dict of entries?

Reply:
In the op case, we're accepting an actual metadata dictionary, here, the supported format is individual metadata entries. Seems to be an API quirk that predates this PR, but I'm open to changing it to just logging metadata directly.

Reply:
I see - yeah, a metadata dictionary would be better. I don't think it occurred to us to allow users to yield a dictionary, but that might have made more sense. I don't think we need to honor the existing API in that way.